bots/tests-data

```diff
@@ -0,0 +1,254 @@
+                        "status": status,
+                        "context": context,
+                        "date": created,
+                        "head": merged,
+                        "test": name,
+                        "url": url,
+                        "log": body
+                    }))
+                    sys.stdout.write("\n")
+            if merged:
+                merged = False
+
+
+# An HTML parser that just pulls out all the <a href="...">
+# link hrefs in a given page of content. We also qualify these
+# hrefs with a base url, in case they're relative
+class HrefParser(HTMLParser.HTMLParser):
+    def __init__(self, base, hrefs):
+        HTMLParser.HTMLParser.__init__(self)
+        self.hrefs = hrefs
+        self.base = base
+
+    def handle_starttag(self, tag, attrs):
+        if tag.lower() == "a":
+            for (name, value) in attrs:
+                if name.lower() == "href":
+                    url = urlparse.urljoin(self.base, value)
+                    # print 'HREF', url
+                    self.hrefs.append(url)
+
+# Retrieves the content of the given URL
+def retrieve(url):
+    return urllib2.urlopen(url).read()
+
+# Returns a list of all results at the given URL
+def links(url):
+    result = [ ]
+    parser = HrefParser(url, result)
+    try:
+        parser.feed(retrieve(url))
+    except urllib2.HTTPError, ex:
+        if ex.code != 404:
+            raise
+    except urllib2.URLError:
+        pass
+    return result
+
+# Generates full GitHub JSON closed pull request
+# objects since the given date. We only care about closed
+# pull requests for processing, the rest are still in flux.
+def pulls(state='closed', since=None):
+    for pull in task.api.pulls(state=state, since=since):
+        yield pull
+
+# Generates revisions for a given pull request. Each revision
+# is a simple string sha. The first one is the one that is at
+# the head of the pull request, and any other follow on revisions
+# are those which were previously tested (eg: force push, or with
+# later fixups).
+def revisions(pull):
+    head = pull["head"]["sha"]
+    if not head:
+        return
+
+    # First give back the main pull request
+    yield head
+
+    # All the revisions we've seen
+    seen = { head: True }
+
+    # Ignore these sink urls. See below. This is an optimization.
+    # We don't need to retrieve the sink /status page for the head revision
+    ignore = { }
+
+    # Seed the set of sinks. We use these sinks to figure out additional
+    # revisions for the pull request. Unfortunately GitHub doesn't help us
+    # with a list of revisions that this pull request used to reflect. So
+    # we have to look to our sink for that info.
+    data = task.api.get("commits/{0}/status?page=1&per_page=100".format(head))
+    for status in data.get("statuses", [ ]):
+        url = status["target_url"]
+        if url:
+            ignore[urlparse.urljoin(url, "./")] = True
```

Comment:
what does this urljoin do?

Reply:
Changed ignore to a set. 

It ignores links back to the current sink output. Since we've already seen it. Added a comment.